### PR TITLE
feature/OP-1308: Build Scripts for Consul

### PR DESCRIPTION
**Pull request by johnyu95**
This branch contains the build scripts for Consul which will be the configuration management tool for OpenRecords. These scripts will create a VM for Consul that is separate from the default machine. Nginx will also be installed on the Consul VM. After the scripts are finished running you need to start a Consul server using the command:

consul agent -config-dir /etc/consul.d/bootstrap

You must do this after the scripts because once a Consul server is started, it will block until an action is taken so the scripts will not proceed once it is started. You can test the installation by accessing the Consul web UI at:

https://10.0.0.5:8501

When running consul_setup.sh you can provide the path to existing Consul binaries as a command line argument for an offline installation:

./consul_setup.sh -f <PATH_TO_CONSUL_BINARIES>

### CODE REVIEW

File: build_scripts/consul_setup/consul_setup.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+# Install Consul binaries
+cd /tmp/
+wget https://releases.hashicorp.com/consul/1.0.0/consul_1.0.0_linux_amd64.zip
+unzip consul_1.0.0_linux_amd64.zip
+sudo mv consul /usr/bin/
```
Comment (joelbcastillo) [inline build_scripts/consul_setup/consul_setup.sh]: Can we also do this in `/usr/local/bin`? System packages are normally installed in `/usr/bin`, user packages in `/usr/local/bin`

**Comment by joelbcastillo** [comment]
@johnyu95 Sorry if I didn't explain myself earlier. We shouldn't be including binary files in our repository, but we should make an option in our build scripts to provide a binary file if we don't have internet access on the server we are installing our application and its dependencies on. In this case, you should remove the binary for consul from the repository and instead use an if statement to check and see if a path to an archive containing a binary has been passed as an argument to the script. If it has, then you should just use the file provided, otherwise, download it from the official release page. 

Does this make sense?

**Comment by johnyu95** [comment]
Oh I see. Okay I'll add it in.

File: build_scripts/consul_setup/config.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+    "bootstrap": true,
+    "server": true,
+    "ui": true,
+    "node_name": "openrecords_consul",
+    "bind_addr": "10.0.0.5",
+    "data_dir": "/var/consul",
+    "log_level": "INFO",
+    "enable_syslog": true,
+    "encrypt": "pmsKacTdVOb4x8/Vtr9PWw==",
```
Comment (joelbcastillo) [inline build_scripts/consul_setup/config.json]: Can we make this key say something like `CHANGE THIS BEFORE DEPLOYING TO PRODUCTION`

I don't want to forget..... 😅

File: build_scripts/consul_setup/consul_setup.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+# 1. Install Consul binaries
+# if the path to the Consul binaries were provided as a command line argument, then copy it to /usr/local/bin
+if [ -n "$1" ]; then
```
Comment (joelbcastillo) [inline build_scripts/consul_setup/consul_setup.sh]: 👍

File: build_scripts/consul_setup/consul_setup.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# 3. Encrypt Consul
+mkdir -p /etc/consul.d/ssl/CA
+chmod 0700 /etc/consul.d/ssl/CA
+cd /etc/consul.d/ssl/CA
+echo "000a" > serial
+ln -s /vagrant/build_scripts/consul_setup/myca.conf /etc/consul.d/ssl/CA/myca.conf
+touch certindex
+
+# Create certificates
+openssl req -x509 -newkey rsa:2048 -days 365 -nodes -out ca.cert -subj "/C=US/ST=New York/L=New York/O=NYC Department of Records and Information Services/OU=IT/CN=consul"
```
Comment (joelbcastillo) [inline build_scripts/consul_setup/consul_setup.sh]: Why are you using your own CA to sign certificates?

**Comment by johnyu95** [comment]
Looks good. Tested locally and consul/vault is working properly.

**Pull request merged**